### PR TITLE
Null converter feature test

### PR DESCRIPTION
**Pull request by Quinn-With-Two-Ns**
Add a test for the null data converter.

I renamed the folder because `null` is a keyword in java.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ Prerequisites:
 - [JDK](https://adoptium.net/?variant=openjdk11&jvmVariant=hotspot) 11+
 - [Node](https://nodejs.org) 16+
 - [Python](https://www.python.org/) 3.10+
-  - [Poetry](https://python-poetry.org/)
+  - [Poetry](https://python-poetry.org/): `poetry install`
```
Comment (bergundy) [inline README.md:28]: `poetry install` will install this project's dependencies (if run in the proper folder), I'd clarify this in the README

File: features/data_converter/empty/feature.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+func CheckResult(ctx context.Context, runner *harness.Runner, run client.WorkflowRun) error {
+	// verify client result is null
+	var result interface{}
+	if err := run.Get(ctx, &result); err != nil {
+		return err
+	}
+	if result != nil {
+		return fmt.Errorf("invalid result: %v", result)
+	}
+	history := runner.Client.GetWorkflowHistory(ctx, run.GetID(), "", false, 0)
```
Comment (bergundy) [inline features/data_converter/empty/feature.go:33]: We could probably abstract this even further and add a method to get the completion event for a workflow (applies to all lang harnesses)

**Review by bergundy — COMMENTED** [review]
I would still like to test the null payload converter in Java and Go, we didn't end up testing those.

File: features/data_converter/empty/feature.go (hunk 2 above)
Comment (cretz) [reply features/data_converter/empty/feature.go:33]: I still am not convinced we need to check payloads. IMO, we only need to check end user behavior. If it comes back as nil, that's fine. (but I understand others disagree w/ me here)

File: features/data_converter/empty/feature.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		if attrs != nil {
+			break
+		}
+	}
+	if attrs == nil {
+		return errors.New("could not locate WorkflowExecutionCompleted event")
+	}
+
+	// verify the workflow execution completed event has no payload
+	runner.Require.Empty(attrs.GetResult().GetPayloads())
```
Comment (cretz) [inline features/data_converter/empty/feature.go:53]: If you use `runner.Require.Empty` here, you might as well use `runner.Require.NotNil` for your result check above.

File: features/data_converter/empty/feature.go (hunk 2 above)
Comment (cretz) [reply features/data_converter/empty/feature.go:33]: I still am not convinced we need to check payloads. IMO, we only need to check end user behavior. If it comes back as nil, that's fine. (but I understand others disagree w/ me here)

File: features/data_converter/empty/feature.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    e = next(
+        e
+        for e in history
+        if e.event_type == EventType.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED
+    )
+    payload = e.workflow_execution_completed_event_attributes.result.payloads[0]
+
+    # load JSON payload from `./payload.json` and compare it to JSON representation of result payload
+    with open(
+        os.path.join(os.path.dirname(runner.feature.file), "payload.json"),
```
Comment (cretz) [inline features/data_converter/empty/feature.py:39]: Is it really worth an entire file just to check empty data and a specific metadata value?

**Review by cretz — APPROVED** [review]
Awesome!

Note, this repo is young, so if you're finding helpers that can be used across tests feel free to add them to harnesses/utils.

File: features/data_converter/empty/feature.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from temporalio.api.common.v1 import Payload
+from temporalio.api.enums.v1 import EventType
+from temporalio.client import WorkflowHandle
+from temporalio.exceptions import ApplicationError
+
+from harness.python.feature import Runner, register_feature
+
+
+@activity.defn
+async def null_activity(input: str) -> None:
```
Comment (bergundy) [inline features/data_converter/empty/feature.py:15]:
```suggestion
async def null_activity(input: Optional[str]) -> None:
```

File: features/data_converter/empty/feature.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	// verify client result is null
+	var result interface{}
+	if err := run.Get(ctx, &result); err != nil {
+		return err
+	}
+	runner.Require.Nil(result)
+
+	// get result payload of ActivityTaskScheduled event from workflow history
+	history := runner.Client.GetWorkflowHistory(ctx, run.GetID(), "", false, 0)
+	event, err := harness.FindEvent(history, func(ev *historyProto.HistoryEvent) bool {
```
Comment (bergundy) [inline features/data_converter/empty/feature.go:51]: Check if err is not nil

File: features/data_converter/empty/feature.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// run a workflow that calls an activity with a null parameter.
+func Workflow(ctx workflow.Context) error {
+	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+		ScheduleToCloseTimeout: 1 * time.Minute,
+	})
+
+	err := workflow.ExecuteActivity(ctx, Activity, nil).Get(ctx, nil)
+	return err
```
Comment (bergundy) [inline features/data_converter/empty/feature.go:30]: nit: can condense this into a single line

File: features/data_converter/empty/feature.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  void activity(String input);
+
+  class Impl implements feature {
+    /**
+     * run a workflow that calls an activity with a null parameter.
+     */
+    @Override
+    public void workflow() {
+      var activities = activities(feature.class, builder -> builder
+              .setScheduleToCloseTimeout(Duration.ofMinutes(1)));
```
Comment (bergundy) [inline features/data_converter/empty/feature.java:34]: We typically use `startToClose`, while schedule to close is fine in this specific case, it's not recommended in most cases where activities should retry indefinitely.
Once a schedule to close timeout is hit the activity will no longer be retried.

The reason I mention this is because this repo has some value as "samples" or "reference" repo where users might look up how the system behaves, we want people to copy our recommended best practices.

File: features/data_converter/empty/feature.py (hunk 4 above)
Comment (bergundy) [reply features/data_converter/empty/feature.py:54]: I think so

Comment (bergundy) [reply features/data_converter/empty/feature.py:54]: If we had cross SDK testing capabilities, I'd say we don't have to test the generated payloads but it's nice to know that null represented the same in all SDKs.

File: features/data_converter/empty/feature.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+from temporalio.exceptions import ApplicationError
+
+from harness.python.feature import Runner, register_feature
+
+
+@activity.defn
+async def null_activity(input: Optional[str]) -> None:
+    # check the null input is serialized correctly
+    if input != None:
+        raise ApplicationError("Activity input should be None")
```
Comment (bergundy) [inline features/data_converter/empty/feature.py:19]: Is this non-retryable?

Comment (bergundy) [reply features/data_converter/empty/feature.py:19]: nit: All of the application failures in this feature _should_ be non-retryable to prevent needless retries.

File: features/data_converter/empty/feature.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import { Feature } from '@temporalio/harness';
+import * as assert from 'assert';
+import { ApplicationFailure } from '@temporalio/common';
+import expectedPayload from './payload.json';
+import * as wf from '@temporalio/workflow';
+
+const activitiesImpl = {
+  async activity(input: any): Promise<void> {
+    if (input != undefined) {
+      throw new ApplicationFailure('Activity input should be undefined', null, true, undefined, undefined);
```
Comment (bergundy) [inline features/data_converter/empty/feature.ts:11]: You can just throw an `Error` for retryable or use the `ApplicationFailure.nonRetryable` factory method instead of the constructor.

File: harness/go/harness/util.go [hunk 11]
```diff
@@ -13,3 +15,17 @@ var RetryDisabled = &temporal.RetryPolicy{MaximumAttempts: 1}
 func AppErrorf(msg string, args ...interface{}) error {
 	return temporal.NewApplicationError(fmt.Sprintf(msg, args...), "SdkFeaturesError")
 }
+
+// Find the first event in the history that meets the condition.
+func FindEvent(history client.HistoryEventIterator, cond func(*historypb.HistoryEvent) bool) (*historypb.HistoryEvent, error) {
```
Comment (bergundy) [inline harness/go/harness/util.go:20]: Nice, now that we have this, can you rewrite the binary test to use it?

**Review by bergundy — APPROVED** [review]
Nice, I didn't have too many comments.
Mostly nits, we can merge after you go over this round of comments AFAIC.

File: harness/go/harness/util.go (hunk 11 above)
Comment (Quinn-With-Two-Ns) [reply harness/go/harness/util.go:20]: yep ill add a commit it do that.

File: features/data_converter/empty/feature.py (hunk 9 above)
Comment (Quinn-With-Two-Ns) [reply features/data_converter/empty/feature.py:19]: no reason for it to be retryable, I'll update

File: features/data_converter/empty/feature.java (hunk 8 above)
Comment (Quinn-With-Two-Ns) [reply features/data_converter/empty/feature.java:34]: will do

File: features/data_converter/empty/feature.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import { Feature } from '@temporalio/harness';
+import * as assert from 'assert';
+import { ApplicationFailure } from '@temporalio/common';
+import expectedPayload from './payload.json';
+import * as wf from '@temporalio/workflow';
+
+const activitiesImpl = {
+  async activity(input: any): Promise<void> {
+    if (input != undefined) {
+      throw ApplicationFailure.nonRetryable('Activity input should be undefined', 'BadResult');
```
Comment (bergundy) [inline features/data_converter/empty/feature.ts:11]: nit: the type is optional and `BadResult` is not really a good name for the error.
Not blocking this PR.

File: features/data_converter/empty/feature.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
+import { JSONToPayload } from '@temporalio/common/lib/proto-utils';
+import { Feature } from '@temporalio/harness';
+import * as assert from 'assert';
+import { ApplicationFailure } from '@temporalio/common';
+import expectedPayload from './payload.json';
+import * as wf from '@temporalio/workflow';
+
+const activitiesImpl = {
+  async activity(input: any): Promise<void> {
```
Comment (bergundy) [inline features/data_converter/empty/feature.ts:9]: nit: This could be an optional parameter or union with undefined or just `undefined`.

```suggestion
  async activity(input?: string): Promise<void> {
```

**Review by bergundy — COMMENTED** [review]
This is good to merge, I had a couple more nits but they can be ignored.
